### PR TITLE
776 store date picker in url

### PR DESCRIPTION
**Pull request by Alastair-L**
### Issue #:
https://github.com/beyondessential/tupaia-backlog/issues/776

### CODE REVIEW

File: packages/web-frontend/src/historyNavigation/utils.js
Comment (Alastair-L) [inline packages/web-frontend/src/historyNavigation/utils.js:104]: How should we handle this case?

File: packages/web-frontend/src/actions.js
Comment (Alastair-L) [inline packages/web-frontend/src/actions.js:141]: This feels really hacky to me...
The reason that I thought creating a new action is necessary:
- We need to wait for the measure hierarchy to be populated before we try to update it
- I only know how to do that in sagas

File: packages/web-frontend/src/historyNavigation/historyMiddleware.js
Comment (Alastair-L) [inline packages/web-frontend/src/historyNavigation/historyMiddleware.js:14]: Organize imports

File: packages/web-frontend/src/historyNavigation/utils.js
Comment (Alastair-L) [inline packages/web-frontend/src/historyNavigation/utils.js:95]: I ended up putting this logic here because it is specific to the url.

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (Alastair-L) [inline packages/web-frontend/src/utils/periodGranularities.js:30]: These formats should probably get chosen by someone else. I've just put place holders in.

File: packages/web-frontend/src/sagas.js
Comment (Alastair-L) [inline packages/web-frontend/src/sagas.js:778]: This whole commit is unrelated, and could/should probably be reverted to save regression. What do you think?

File: packages/web-frontend/src/historyNavigation/historyMiddleware.js
Comment (kael89) [reply packages/web-frontend/src/historyNavigation/historyMiddleware.js:14]: Nice!

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (kael89) [reply packages/web-frontend/src/utils/periodGranularities.js:30]: Fair enough! Let's add some information in the card about how the urls will look like, and then we can all decide

File: packages/web-frontend/src/sagas.js
Comment (kael89) [reply packages/web-frontend/src/sagas.js:778]: Before I review this change, could you give me some context about it? Is it intended as pure refactoring?

**Comment by kael89** [comment]
@Alastair-L I will continue reviewing after https://github.com/beyondessential/tupaia/pull/1275#discussion_r485242554 is answered

File: packages/web-frontend/src/sagas.js
Comment (Alastair-L) [inline packages/web-frontend/src/sagas.js]: Should have mentioned earlier, but this section is not part of the refactor (should be with later commits)

File: packages/web-frontend/src/sagas.js
Comment (Alastair-L) [reply packages/web-frontend/src/sagas.js:778]: Yes, it should be purely refactoring (plus a small performance loss? from using selectors (because of the overhead and lack of recalculation in real life))

Comment (Alastair-L) [reply packages/web-frontend/src/sagas.js:778]: The motivation was that I found the existing code very confusing.

Comment (kael89) [reply packages/web-frontend/src/sagas.js:778]: Fair enough! I think it is a valuable refactor. Remember to leave regression testing notes in the card

Comment (Alastair-L) [reply packages/web-frontend/src/sagas.js:778]: Will do 👍

File: packages/web-frontend/src/historyNavigation/utils.js
Comment (kael89) [reply packages/web-frontend/src/historyNavigation/utils.js:117]: As discussed, while this may indicate an error (start and end date being different for a one-date granularity), it's not our concern here. So we could as well as remove the check and error logging, and leave this concern to places where this state of data does matter

File: packages/web-frontend/src/actions.js
Comment (kael89) [reply packages/web-frontend/src/actions.js:141]: I'll admit my ignorance regarding sagas that must run in sync, @rohan-bes any suggestions for a clean approach here?

File: packages/web-frontend/src/actions.js
Comment (kael89) [inline packages/web-frontend/src/actions.js:564]:
```suggestion
export function updateMeasureDateRangeOnceHierarchyLoads(periodString) {
```

(can also keep `Current` in the name if you thing it's required)

File: packages/web-frontend/src/actions.js
Comment (kael89) [reply packages/web-frontend/src/actions.js:141]: But if we keep this, I suggest renaming it to `UPDATE_MEASURE_DATE_RANGE_ONCE_HIERARCHY_LOADS` to make our intentions clearer

**Review by kael89 — CHANGES_REQUESTED** [review]
Looks good overall! A couple of renames requested, and waiting for feedback on `UPDATE_MEASURE_CONFIG_ONCE_HIERARCHY_LOADS `

Could you add unit tests for the `periodString <=> dateRange` conversion utils?

Also, I hope that import reordering will not cause massive conflicts in other PRs. If it does it should be easy to solve those by also ordering the imports in the other PR, and then comparing. At any case it may worth letting people know in the `dev` channel after this is merged into `dev`

File: packages/web-frontend/src/historyNavigation/utils.js
Comment (Alastair-L) [reply packages/web-frontend/src/historyNavigation/utils.js:95]: `DateRange` is a good name 👍

File: packages/web-frontend/src/actions.js
Comment (rohan-bes) [reply packages/web-frontend/src/actions.js]: Didn't get a chance to properly look into this one unfortunately, but I can't immediately see a better way of doing it. I'm cool with this going in as it is for now, we always have the option to refactor later.

Comment (kael89) [reply packages/web-frontend/src/actions.js]: Thanks @rohan-bes !

File: packages/web-frontend/src/tests/historyNavigation/UrlPeriodStringToDateRangeConversions.test.js
Comment (kael89) [inline packages/web-frontend/src/tests/historyNavigation/UrlPeriodStringToDateRangeConversions.test.js:29]: Using `isSame` is clever and clean! 💯

Can I suggest an optional change: it took me a bit to realise that `isSame` comes from `moment` and not from jest (although at a closer look, the syntax does not imply that). You can add a `assertMomentDatesAreEqual ` helper:

```js
const assertMomentDatesAreEqual = (moment1, moment2) => expect(moment1.isSame(moment2)).toBe(true)`
```

**Review by kael89 — CHANGES_REQUESTED** [review]
Nice tests! One requested and one optional change.

By the way, we still need to decide on the date formats... Given that your code is easily configurable (+ we have tests), I happy for you to merge it in `url-refactor` and then change it if a different format is required. Could you ask a question in the card about the format?

File: packages/web-frontend/src/tests/historyNavigation/UrlPeriodStringToDateRangeConversions.test.js
Comment (Alastair-L) [reply packages/web-frontend/src/tests/historyNavigation/UrlPeriodStringToDateRangeConversions.test.js:29]: Clean!

File: packages/web-frontend/src/tests/historyNavigation/UrlPeriodStringToDateRangeConversions.test.js
Comment (Alastair-L) [reply packages/web-frontend/src/tests/historyNavigation/UrlPeriodStringToDateRangeConversions.test.js:7]: Good point.

**Comment by Alastair-L** [comment]
Fyi, I want to wait until `url-refactor` is merged to dev before this card is tested, as the priority is getting the testing on that finished I believe, which Bindu is almost finished.

**Review by kael89 — APPROVED** [review]

**Comment by Alastair-L** [comment]
^ Force pushed due to the squash and merge of the url-refactor 😢

**Comment by kael89** [comment]
> ^ Force pushed due to the squash and merge of the url-refactor 😢

No worries at this case!

**Comment by kael89** [comment]
@rohan-bes the same `OneTimeLogin` flaky test failed in this build... 🤔

File: packages/web-frontend/src/tests/components/dateRangePicker.test.js
Comment (Alastair-L) [inline packages/web-frontend/src/tests/components/dateRangePicker.test.js:30]: Is it ok to use this util as the util itself is not being tested here?

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (Alastair-L) [inline packages/web-frontend/src/utils/periodGranularities.js:129]: Hurray for unit tests!

File: packages/web-frontend/src/tests/components/dateRangePicker.test.js
Comment (kael89) [reply packages/web-frontend/src/tests/components/dateRangePicker.test.js:30]: It's actually a better practice to use the end value directly

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (kael89) [reply packages/web-frontend/src/utils/periodGranularities.js:129]: Oh man it would be so hard to catch this if not for unit tests... Mutability is one of the key points of moment.js criticism

File: packages/web-frontend/src/tests/components/dateRangePicker.test.js
Comment (Alastair-L) [reply packages/web-frontend/src/tests/components/dateRangePicker.test.js:30]: Should I create a map up the top?

```javascript
MIN_MOMENT_STRINGS = {
  [GRANULARITIES.SINGLE_DAY]: '20 Jan 2020',
  ...
};
```
Then 
```javascript
const startDate = MIN_MOMENT_STRINGS[key];
```

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (Alastair-L) [reply packages/web-frontend/src/utils/periodGranularities.js:129]: Yeah I should have checked that `.startOf` didn't mutate the original, but I just assumed it wouldn't 😬. Three cheers for https://github.com/beyondessential/tupaia-backlog/issues/1303

File: packages/web-frontend/src/tests/components/dateRangePicker.test.js
Comment (kael89) [reply packages/web-frontend/src/tests/components/dateRangePicker.test.js:30]: Yeap sounds good!

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (kael89) [inline packages/web-frontend/src/utils/periodGranularities.js:123]: I find `{{typeA}}to{{typeB}}`/`{{formatA}}to{{formatB}}` names quite intuitive for this sort of functionality:

```suggestion
export const momentToDateString = (date, granularity, format) =>
```

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (kael89) [inline packages/web-frontend/src/utils/periodGranularities.js:124]: Let's use the symbolic constants defined in this file instead of `week`/`one_week_at_at_time`

**Review by kael89 — CHANGES_REQUESTED** [review]
Looks good overall! A couple of changes and it's ready to go

**Comment by kael89** [comment]
@Alastair-L you'll also need to resolve conflicts

**Comment by Alastair-L** [comment]
Thanks for the review @kael89. I'll get another PR up and then come back to this.

File: packages/web-frontend/src/utils/periodGranularities.js
Comment (Alastair-L) [reply packages/web-frontend/src/utils/periodGranularities.js:123]: Me too now that I see it!

File: packages/web-frontend/src/tests/components/dateRangePicker.test.js
Comment (Alastair-L) [inline packages/web-frontend/src/tests/components/dateRangePicker.test.js:65]: This is more so that if it changes and this test fails, the error is easier to diagnose, as opposed to testing any functionality.

Comment (kael89) [reply packages/web-frontend/src/tests/components/dateRangePicker.test.js:65]: Good idea!

**Review by kael89 — APPROVED** [review]
Looks good!

**Comment by Alastair-L** [comment]
Can I merge this, as it was testing passed here: https://github.com/beyondessential/tupaia/pull/1275#issuecomment-696518759, and since I have only merged dev and made those small changes? 

I will give it a smoke test after it builds in dev.

What do you think?

**Comment by kael89** [comment]
@Alastair-L sounds good!

**Pull request merged**
